examples/cis2-wccd-upgradable/src/lib.rs

```diff
@@ -0,0 +1,3181 @@
+        state_address:          params.state_address,
+        implementation_address: params.implementation_address,
+    };
+
+    Ok(state)
+}
+
+/// Initializes the state of the w_ccd contract with the proxy and the
+/// implementation addresses. Both addresses have to be set together by calling
+/// this function. This function can only be called once.
+#[receive(
+    contract = "CIS2-wCCD-State",
+    name = "initialize",
+    parameter = "InitializeStateParams",
+    mutable
+)]
+fn contract_state_initialize<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+) -> ContractResult<()> {
+    ensure!(
+        host.state().proxy_address.is_none() && host.state().implementation_address.is_none(),
+        concordium_cis2::Cis2Error::Custom(CustomContractError::AlreadyInitialized)
+    );
+    // Set proxy and implementation addresses.
+    let params: InitializeStateParams = ctx.parameter_cursor().get()?;
+    host.state_mut().proxy_address = params.proxy_address;
+    host.state_mut().implementation_address = params.implementation_address;
+    Ok(())
+}
+
+/// Initializes the implementation of the w_ccd contract with the proxy and the
+/// state addresses. Both addresses have to be set together by calling this
+/// function. This function can only be called once.
+#[receive(
+    contract = "CIS2-wCCD",
+    name = "initialize",
+    parameter = "InitializeImplementationParams",
+    mutable
+)]
+fn contract_initialize<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<StateImplementation, StateApiType = S>,
+) -> ContractResult<()> {
+    ensure!(
+        host.state().proxy_address.is_none() && host.state().state_address.is_none(),
+        concordium_cis2::Cis2Error::Custom(CustomContractError::AlreadyInitialized)
+    );
+    // Set proxy and storage addresses.
+    let params: InitializeImplementationParams = ctx.parameter_cursor().get()?;
+    host.state_mut().proxy_address = params.proxy_address;
+    host.state_mut().state_address = params.state_address;
+    Ok(())
+}
+
+/// The fallback method, which redirects the invocations to the implementation.
+#[receive(contract = "CIS2-wCCD-Proxy", fallback, mutable, payable)]
+fn receive_fallback<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<StateProxy, StateApiType = S>,
+    amount: Amount,
+) -> ReceiveResult<RawReturnValue> {
+    let entrypoint = ctx.named_entrypoint();
+    let implementation = host.state().implementation_address;
+    let mut parameter_buffer = vec![0; ctx.parameter_cursor().size() as usize];
+    ctx.parameter_cursor().read_exact(&mut parameter_buffer)?;
+
+    // The `proxy` will set the `current_sender` on the `implementation` before
+    // forwarding the invoke unaltered to the implementation contract.
+    let parameter_bytes = to_bytes(&SetCurrentSenderParams {
+        sender: ctx.sender(),
+    });
```

Comment:
My initial proposal was to append the `sender` to the `parameter_buffer`. Looking at the code now I was not exactly sure how to append this buffer in Rust especially because we need to append it by an `Address` which has a different format/size when it is a `Contract` address compared to when it is an `Account`. Maybe that is not a problem at all to append it by an `Address` but I haven’t tried it yet.

The current approach has an advantage in terms of schema generation. I am able to use the same schema type from the implementation and add it to the fallback function on the proxy. E.g. I use “WrapParams” on the fallback function on the proxy when I want to execute the “wrap” function in the protocol.

Appending an address to the buffer seems cheaper while the other approach might make schema integration easier. I assume, we need to discuss this a bit and then choose the better approach.